tl.tl

```diff
@@ -13757,22 +13756,42 @@ do
                   end
                   t = u
 
+               -- prefer the union type if it already works and the other one isn't a union
+               elseif ra is UnionType and not (rb is UnionType) and self:is_a(rb, ra) then
+                  -- convert it to the union type
+                  t = drop_constant_value(ra)
+
+               elseif rb is UnionType and not (ra is UnionType) and self:is_a(ra, rb) then
+                  -- convert it to the union type
+                  t = drop_constant_value(rb)
+
                else
-                  local a_ge_b = self:is_a(rb, ra)
-                  local b_ge_a = self:is_a(ra, rb)
+                  local a_ge_b = self:is_a(ub, ua)
+                  local b_ge_a = self:is_a(ua, ub)
                   node.known = facts_or(node, node.e1.known, node.e2.known)
-                  if a_ge_b or b_ge_a then
-                     if expected then
-                        local a_is = self:is_a(ua, expected)
-                        local b_is = self:is_a(ub, expected)
-                        if a_is and b_is then
-                           t = self:infer_at(node, expected)
-                        end
-                     end
-                     if not t then
-                        local larger_type = b_ge_a and ub or ua
-                        t = larger_type
-                     end
+                  local is_same = self:same_type(ra, rb)
```

Comment:
Should this be ua and ub like above or does that not work?

Reply:
@hishamhm 
I've tested with using unresolved types and it found another couple of mistakes in the compiler. I did try it, but it then breaks table statements with `or` etc. I'd think just leave it as is for the moment and it can be changed to the unresolved types once we have some better inferencing for `table -> nominal`.